frame/support/src/traits/hooks.rs

```diff
@@ -18,30 +18,34 @@
 //! Traits for hooking tasks to events in a blockchain's lifecycle.
 
 use impl_trait_for_tuples::impl_for_tuples;
-use sp_arithmetic::traits::Saturating;
+use sp_arithmetic::traits::{Saturating, Zero};
 use sp_runtime::traits::AtLeast32BitUnsigned;
 
 /// The block initialization trait.
 ///
 /// Implementing this lets you express what should happen for your pallet when the block is
 /// beginning (right before the first extrinsic is executed).
-pub trait OnInitialize<BlockNumber> {
+pub trait OnInitialize<BlockNumber, Weight: Zero> {
```

Comment:
Here's one thing that we can consider -- is it better to have `Weight` as an associated type, rather than a type parameter to a trait? The difference here is that an associated type is defined by the _implementer_ of the trait, rather than the _caller_ or _user_ of the trait.

This would allow you to omit the `Weight` type parameter in cases where we need to typecast something into `OnInitialize` or `Hooks`.

Reply:
I am just following the pattern already set out by `BlockNumber` here. Also there seems to be issues trying to make this an associated type and having the `impl_trait_for_tuples` work.